app/controllers/password_resets_controller.rb

```diff
@@ -0,0 +1,23 @@
+class PasswordResetsController < ApplicationController
+  def show
+    user = User.find_by(token: params[:id])
+    if user
+      @token = user.token
+    else
+      redirect_to expired_token_path unless user
+    end
+  end
+
+  def create
+    user = User.find_by(token: params[:token])
+    if user
+      user.password = params[:password]
+      user.save!
+      user.update_with_token!
```

Comment:
These last 4 lines are not correct.

```ruby
if user
  user.password = params[:password]
  user.save!
  user.update_with_token!
```

* First reason why not: if user enters password that is not valid, you need to return back to him and not blow it with exception `save!`
* Second reason is that when you set an attribute and save right away you should just use `update` straight.
* Third reason is - instead of using `update_with_token!` you should not create another token, but blow it away - remove the token once its used, otherwise its not a better solution that the previous one.

This is how it should look like:

```ruby
if user && user.update(password: params[:password], token: nil)
  flash[:success] = "Your password has been changed. Please sign in."
  redirect_to login_path
elsif user
  @token = user.token
  render :show
else
  redirect_to expired_token_path
end
```